src/AppBundle/Service/ReportService.php

```diff
@@ -296,4 +324,84 @@ public function findDeleteableReports(Collection $reports)
 
         return $ret;
     }
+
+
+    /**
+     * @param string $select reports|count
+     * @param string $status see Report::STATUS_* constants
+     * @param integer $userId
+     * @param boolean $exclude_submitted
+     * @param string $q search query client firstname/lastname or case number
+     *
+     * @return QueryBuilder
+     */
+    public function getAllReportsQb($select, $status, $userId, $exclude_submitted, $q)
+    {
+        $qb = $this->_em->getRepository(Report::class)
+            ->createQueryBuilder('r');
+
+        if ($select == 'reports') {
+            $qb
+                ->select('r,c,u')
+                ->leftJoin('r.submittedBy', 'sb');
+        } elseif ($select == 'count') {
+            $qb->select('COUNT(r)');
+        } else {
+            throw new \InvalidArgumentException(__METHOD__ . ": first must be reports|count");
+        }
+
+        $qb
+            ->leftJoin('r.client', 'c')
+            ->leftJoin('c.users', 'u')
+            ->where('u.id = ' . $userId);
+
+        if ($exclude_submitted) {
+            $qb->andWhere('r.submitted = false OR r.submitted is null');
+        }
+
+        if ($q) {
+            $qb->andWhere('lower(c.firstname) LIKE :qLike OR lower(c.lastname) LIKE :qLike OR c.caseNumber = :q');
+            $qb->setParameter('qLike', '%' . strtolower($q) . '%');
+            $qb->setParameter('q', $q);
+        }
+
+        // note: reportStatusCached is stored ignoring due date
+        if ($status == Report::STATUS_READY_TO_SUBMIT) {
+            // reports ready to submit are when reportStatusCached=readyToSubmit AND is also due (enddate < today)
+            $qb->andWhere('r.reportStatusCached = :status AND r.endDate < :endOfToday')
+                ->setParameter('status', $status)
+                ->setParameter('endOfToday', $this->endOfToday);
+        } else if ($status == Report::STATUS_NOT_FINISHED) {
+            // report not finished are report with reportStatusCached=notFinished
+            // OR ready to submit but not yet due
+            $qb->andWhere('r.reportStatusCached = :status OR (r.reportStatusCached = :readyToSubmit AND r.endDate >= :endOfToday)')
+                ->setParameter('status', $status)
+                ->setParameter('readyToSubmit', Report::STATUS_READY_TO_SUBMIT)
+                ->setParameter('endOfToday', $this->endOfToday);
+        } else if ($status == Report::STATUS_NOT_STARTED) {
+            $qb->andWhere('r.reportStatusCached = :status')
+                ->setParameter('status', $status);
+        }
+
+        return $qb;
+    }
+
+    /**
+     * If the report is ready to submit, but is not yet due, return notFinished instead
+     * In all the the cases, return original $status
+     *
+     * @param $status
+     * @param \DateTime $endDate
+     *
+     * @return string
+     */
+    public function adjustReportStatus($status, \DateTime $endDate)
```

Comment:
This method shouldnt be needed if reportStatus object is doing its job properly?

Reply:
it's needed to be called after the array hydration. The reportStatusCache does not store the dueDate, but the client expects the report status to be readyToSubmit if also the date is due.